compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -251,6 +281,314 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
         }
     }
 
+    fn insert_scalar(&mut self, scalar: Scalar, ty: Ty<'tcx>) -> VnIndex {
+        self.insert(Value::Constant(Const::from_scalar(self.tcx, scalar, ty)))
+    }
+
+    #[instrument(level = "trace", skip(self), ret)]
+    fn eval_to_const(&mut self, value: VnIndex) -> Option<OpTy<'tcx>> {
+        use Value::*;
+        let op = match *self.get(value) {
+            Opaque(_) => return None,
+            // Do not bother evaluating repeat expressions. This would uselessly consume memory.
+            Repeat(..) => return None,
+
+            Constant(ref constant) => self.ecx.eval_mir_constant(constant, None, None).ok()?,
+            Aggregate(kind, variant, ref fields) => {
+                let fields = fields
+                    .iter()
+                    .map(|&f| self.evaluated[f].as_ref())
+                    .collect::<Option<Vec<_>>>()?;
+                let ty = match kind {
+                    AggregateTy::Array => {
+                        assert!(fields.len() > 0);
+                        Ty::new_array(self.tcx, fields[0].layout.ty, fields.len() as u64)
+                    }
+                    AggregateTy::Tuple => {
+                        Ty::new_tup_from_iter(self.tcx, fields.iter().map(|f| f.layout.ty))
+                    }
+                    AggregateTy::Def(def_id, args) => {
+                        self.tcx.type_of(def_id).instantiate(self.tcx, args)
+                    }
+                };
+                let variant = if ty.is_enum() { Some(variant) } else { None };
+                let ty = self.ecx.layout_of(ty).ok()?;
+                if ty.is_zst() {
+                    ImmTy::uninit(ty).into()
+                } else if matches!(ty.abi, Abi::Scalar(..) | Abi::ScalarPair(..)) {
+                    let alloc_id = self
+                        .ecx
+                        .intern_with_temp_alloc(ty, |ecx, dest| {
+                            let variant_dest = if let Some(variant) = variant {
+                                ecx.project_downcast(dest, variant)?
+                            } else {
+                                dest.clone()
+                            };
+                            for (field_index, op) in fields.into_iter().enumerate() {
+                                let field_dest = ecx.project_field(&variant_dest, field_index)?;
+                                ecx.copy_op(op, &field_dest, /*allow_transmute*/ false)?;
+                            }
+                            ecx.write_discriminant(variant.unwrap_or(FIRST_VARIANT), dest)
+                        })
+                        .ok()?;
+                    let mplace =
+                        self.ecx.raw_const_to_mplace(ConstAlloc { alloc_id, ty: ty.ty }).ok()?;
+                    mplace.into()
+                } else {
+                    return None;
+                }
+            }
```

Comment:
@RalfJung those are the 2 cases where we need to allocate some memory for the interpreter.
`eval_mir_constant` already reuses the alloc-id stored in the `ConstValue`, or makes a new pointer to global memory.
As for aggregates, I directly intern the allocation for this purpose.

I can add an assertion at the end of `eval_to_const` to verify that if the `OpTy` is indirect, it points to global memory.

Reply:
I don't think I understand the relevant invariant here. I thought this was about preserving pointer equalities? `eval_mir_constant` will create new `AllocId` in some cases, and `intern_with_temp_alloc` always creates a new `AllocId`, so both of these will create fresh inequal pointers. Isn't that the problem you wanted to avoid?

Reply:
The problem I wanted to avoid was dealing with mutability of the allocations. By using `intern_with_temp_alloc`, once a value is written, I'm certain it won't be modified accidentally.

Reply:
Oh, I completely misunderstood then.

You can mark allocation as immutable even without interning them into the global state, and the interpreter will respect that. But of course things you want to insert into MR need to be in the global alloc map. So the question is, how often are you causing global allocs to be created here that won't be used by the eventually produced MR? If that number is big that's a waste of memory and it'd be better to use the local memory of the `InterpCx` instead, and only intern at the end what is actually used.

Reply:
Changed to use `allocate` and `alloc_mark_immutable`.